ceno_zkvm/src/structs.rs

```diff
@@ -273,10 +275,18 @@ impl<E: ExtensionField> ZKVMWitnesses<E> {
             self.combined_lk_mlt.as_ref().unwrap(),
             input,
         )?;
-        assert!(self.witnesses.insert(TC::name(), witness).is_none());
+        assert!(self.witnesses_tables.insert(TC::name(), witness).is_none());
+        assert!(!self.witnesses_opcodes.contains_key(&TC::name()));
 
         Ok(())
     }
+
+    /// Iterate opcode circuits, then table circuits, sorted by name.
+    pub fn into_iter_sorted(self) -> impl Iterator<Item = (String, RowMajorMatrix<E::BaseField>)> {
+        self.witnesses_opcodes
```

Comment:
Cosmetic suggestion: use itertools, we already have it as a dependency anyway.

```diff
diff --git a/ceno_zkvm/src/structs.rs b/ceno_zkvm/src/structs.rs
index b29aba88..5d91eaad 100644
--- a/ceno_zkvm/src/structs.rs
+++ b/ceno_zkvm/src/structs.rs
@@ -9,7 +9,7 @@ use crate::{
 };
 use ceno_emul::StepRecord;
 use ff_ext::ExtensionField;
-use itertools::Itertools;
+use itertools::{chain, Itertools};
 use mpcs::PolynomialCommitmentScheme;
 use multilinear_extensions::{
     mle::DenseMultilinearExtension, virtual_poly_v2::ArcMultilinearExtension,
@@ -283,9 +283,10 @@ impl<E: ExtensionField> ZKVMWitnesses<E> {
 
     /// Iterate opcode circuits, then table circuits, sorted by name.
     pub fn into_iter_sorted(self) -> impl Iterator<Item = (String, RowMajorMatrix<E::BaseField>)> {
-        self.witnesses_opcodes
-            .into_iter()
-            .chain(self.witnesses_tables)
+        chain(
+            self.witnesses_opcodes,
+            self.witnesses_tables,
+        )
     }
 }
```

Reply:
Done.